wolfcrypt/src/aes.c

```diff
@@ -2367,8 +2414,14 @@ static WARN_UNUSED_RESULT WC_INLINE word32 PreFetchTd4(void)
 #endif
 
 /* Software AES - ECB Decrypt */
+#ifdef NEED_AES_HW_FALLBACK
```

Comment:
can we follow the pattern of AESNI and FREESCALE_LTC here where only the existing NEED_AES_TABLES macro is needed

Reply:
I'm not sure I understand. Do you mean to use `NEED_AES_TABLES` _instead of_  `NEED_AES_HW_FALLBACK`?

Let me clarify my intentions. The issue here is _only_ with the ESP32-S3, and not the classic ESP32.  

The classic ESP32 has all AES algos implemented in hardware. I see how that could follow the FREESCALE_LTC pattern.

However, the  ESP32-S3 only has _partial_ AES hardware acceleration implementation. It is missing AES-192.  See https://github.com/wolfSSL/wolfssl/issues/6375

Thus for the ESP32, it is all or nothing: either we need AES tables for SW, or not when we use HW.

For the ESP32-S3, we _may_ need tables. So depending on the specific AES (e.g. [AES192 enabled](https://github.com/gojimmypi/wolfssl/blob/c18fb50744dbf35bbc49fa5ffe0706182a691334/wolfcrypt/src/aes.c#L683)), we need both a hardware and software algo enabled. The `NEED_AES_HW_FALLBACK` is used to indicate this.

For example, [here](https://github.com/gojimmypi/wolfssl/blob/c18fb50744dbf35bbc49fa5ffe0706182a691334/wolfcrypt/src/aes.c#L725) we check `NEED_AES_HW_FALLBACK` and if enabled, we check if `aes` parameter is a supported key length for hardware, otherwise fall back to software:

```
    #ifdef NEED_AES_HW_FALLBACK
        if (wc_esp32AesSupportedKeyLen(aes)) {
            ret = wc_esp32AesDecrypt(aes, inBlock, outBlock);
        }
        else {
            ret = wc_AesDecrypt_SW(aes, inBlock, outBlock);
        }
    #else
        /* if we don't need fallback, always use HW */
        ret = wc_esp32AesDecrypt(aes, inBlock, outBlock);
    #endif
``` 

If that was not clear, perhaps I could use a more intuitive name?

It could be I completely misunderstood. :)  If so, please clarify. Thanks

Reply:
Am hoping we can get away from having the ifdef around the function name. Either by having the esp function calls instead be in the wc_AesDecrypt() function or by following the wc_AesGcmDecrypt of calling a AES_GCM_decrypt_C function like with AESNI.

Reply:
ok, I've pushed some of the requested changes, but the AES update was a bit more complicated. 

My [WIP is over on my dev branch](https://github.com/gojimmypi/wolfssl/commits/ED25519_SHA2_fix/wolfcrypt/src/aes.c). I did get rid of the `ifdef` around the function name, for example [here](https://github.com/gojimmypi/wolfssl/blob/7378e9fd6ed651b2a669f9bd69353fcf672ac01a/wolfcrypt/src/aes.c#L4267):

```
    #if defined(WOLFSSL_ESPIDF) && defined(NEED_AESCBC_HW_FALLBACK)
        if (wc_esp32AesSupportedKeyLen(aes)) {
            ESP_LOGV(TAG, "wc_AesCbcEncrypt calling wc_esp32AesCbcEncrypt");
            return wc_esp32AesCbcEncrypt(aes, out, in, sz);
        }
        else {
            ESP_LOGW(TAG, "wc_AesCbcEncrypt unsupported keylen = %d, "
                          "falling back to SW.", aes->keylen);
        }
    #endif

software algo follows...
``` 

It needs some more polishing and testing. I expect to be able to update this PR in the coming week. cc: @PaulMartinsen

Reply:
Hi @JacobBarthelmeh, I pushed the [updates](https://github.com/wolfSSL/wolfssl/pull/6624/commits/d0756bd406518b7e6ad812d1c076df88d7b6c474) that should address the last of the code review comments. In particular I'm no longer renaming function for software fallback.

Let me know how things are looking. Thank you.